bots/tests-data

```diff
@@ -0,0 +1,254 @@
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import json
+import os
+import sys
+import subprocess
+import urlparse
+import urllib2
+
+import HTMLParser
+
+sys.dont_write_bytecode = True
+
+import task
+
+BOTS = os.path.abspath(os.path.dirname(__file__))
+SINKS = { }
+
+def main():
+    parser = argparse.ArgumentParser(description="Pull out test data for pull requests")
+    parser.add_argument("--open", action="store_true", help="Pull data on open pull requests")
+    parser.add_argument("--since", nargs="?", help="Since a given pull request or date")
+    parser.add_argument("-v", "--verbose", action="store_true", help="Show verbose progress")
+    opts = parser.parse_args()
+
+    for pull in pulls(opts.open and "open" or "closed", since=opts.since):
+        if opts.open:
+            merged = None
+        else:
+            merged = True
+        if opts.verbose:
+            sys.stderr.write("pull-{0}\n".format(pull["number"]))
+        for revision in revisions(pull):
+            if opts.verbose:
+                sys.stderr.write("- {0}\n".format(revision))
+            for (context, created, url, log) in logs(revision):
+                if opts.verbose:
+                    sys.stderr.write("  - {0} {1}\n".format(created, context))
+                for (status, name, body) in tap(log):
+                    sys.stdout.write(json.dumps({
+                        "pull": pull["number"],
+                        "revision": revision,
+                        "status": status,
+                        "context": context,
+                        "date": created,
+                        "head": merged,
+                        "test": name,
+                        "url": url,
+                        "log": body
+                    }))
+                    sys.stdout.write("\n")
+            if merged:
+                merged = False
+
+
+# An HTML parser that just pulls out all the <a href="...">
+# link hrefs in a given page of content. We also qualify these
+# hrefs with a base url, in case they're relative
+class HrefParser(HTMLParser.HTMLParser):
+    def __init__(self, base, hrefs):
+        HTMLParser.HTMLParser.__init__(self)
+        self.hrefs = hrefs
+        self.base = base
+
+    def handle_starttag(self, tag, attrs):
+        if tag.lower() == "a":
+            for (name, value) in attrs:
+                if name.lower() == "href":
+                    url = urlparse.urljoin(self.base, value)
+                    # print 'HREF', url
+                    self.hrefs.append(url)
```

Comment:
Do duplicates matter/hurt here? You might consider using `self.hrefs.add(url)` instead and initializing it as `set()` instead of `[]` below. As you only iterate over the links with `for`, both lists and sets work fine.

Reply:
I don't think they matter. Since hrefs transfers across classes I'd rather keep it simple.